tmpauthenticator/__init__.py

```diff
@@ -27,26 +28,44 @@ async def get(self):
         login mechanism. This only happens when /tmplogin is hit - so you can use
         other parts of the hub as you normally would.
         """
+        # Create a new user.
+        #
+        # user_from_username ref: https://github.com/jupyterhub/jupyterhub/blob/4.0.0/jupyterhub/handlers/base.py#L504-L505
+        #
         username = str(uuid.uuid4())
         user = self.user_from_username(username)
-        user = await gen.maybe_future(self.process_user(user, self))
-        self.set_login_cookie(user)
 
-        # This sets a hub login cookie for the new user, overwriting old user's cookies if needed
+        # Let a subclasses of TmpAuthenticator process the new user by
+        # overriding TmpAuthenticator.process_user.
+        #
+        user = await maybe_future(self.process_user(user, self))
```

Comment:
I've learned since writing some of this compatibility in JupyterHub that I think a slightly better pattern here is to use `inspect.isawaitable`. That avoids an `await` if it is actually sync:

```suggestion
user = self.process_user(user, self)
if inspect.isawaitable(user):
    user = await user
```

The `maybe_future` was more useful in the transitional days where we needed to accept concurrent Futures, tornado Futures, tornado coroutine objects, asyncio coroutine objects, etc. But now that we're all-in on asyncio, this is a simpler check.

Reply:
I forgot to add - this is a minor stylistic choice and has negligible practical effect. But it does avoid unnecessary awaits. Feel free to merge with or without the suggestion.